ext/mbstring/libmbfl/filters/mbfilter_utf8.c

```diff
@@ -355,3 +580,536 @@ static zend_string* mb_cut_utf8(unsigned char *str, size_t from, size_t len, uns
+{
+	unsigned char *p = *in, *e = p + *in_len;
+	uint32_t *out = buf, *limit = buf + bufsize - 1;
+
+	while (p < e && out < limit) {
+		unsigned char c = *p++;
+		unsigned int s = 0;
+
+		if (c <= 0x7F) {
+			*out++ = c;
+			continue;
+		} else if (c >= 0xC2 && c <= 0xDF && p < e) {
+			unsigned char c2 = *p++;
+
+			if ((c2 & 0xC0) == 0x80) {
+				s = ((c & 0x1F) << 6) | (c2 & 0x3F);
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+				p--;
+				continue;
+			}
+		} else if (c >= 0xE0 && c <= 0xEF) {
+			if ((e - p) < 2) {
+				*out++ = MBFL_BAD_INPUT;
+				if (p < e && (c != 0xE0 || *p >= 0xA0) && (c != 0xED || *p < 0xA0) && (*p & 0xC0) == 0x80) {
+					p++;
+					if (p < e && (*p & 0xC0) == 0x80) {
+						p++;
+					}
+				}
+				continue;
+			}
+			unsigned char c2 = *p++;
+			unsigned char c3 = *p++;
+
+			if ((c2 & 0xC0) != 0x80 || (c == 0xE0 && c2 < 0xA0) || (c == 0xED && c2 >= 0xA0)) {
+				*out++ = MBFL_BAD_INPUT;
+				p -= 2;
+				continue;
+			} else if ((c3 & 0xC0) != 0x80) {
+				*out++ = MBFL_BAD_INPUT;
+				p--;
+				continue;
+			} else {
+				s = ((c & 0xF) << 12) | ((c2 & 0x3F) << 6) | (c3 & 0x3F);
+			}
+		} else if (c >= 0xF0 && c <= 0xF4) {
+			if ((e - p) < 3) {
+				*out++ = MBFL_BAD_INPUT;
+				if (p < e) {
+					unsigned char c2 = *p;
+					if ((c == 0xF0 && c2 >= 0x90) || (c == 0xF4 && c2 < 0x90) || (c >= 0xF1 && c <= 0xF3)) {
+						while (p < e && (*p & 0xC0) == 0x80) {
+							p++;
+						}
+					}
+				}
+				continue;
+			}
+			unsigned char c2 = *p++;
+			unsigned char c3 = *p++;
+			unsigned char c4 = *p++;
+
+			if ((c2 & 0xC0) != 0x80 || (c == 0xF0 && c2 < 0x90) || (c == 0xF4 && c2 >= 0x90)) {
+				*out++ = MBFL_BAD_INPUT;
+				p -= 3;
+				continue;
+			} else if ((c3 & 0xC0) != 0x80) {
+				*out++ = MBFL_BAD_INPUT;
+				p -= 2;
+				continue;
+			} else if ((c4 & 0xC0) != 0x80) {
+				*out++ = MBFL_BAD_INPUT;
+				p--;
+				continue;
+			} else {
+				s = ((c & 0x7) << 18) | ((c2 & 0x3F) << 12) | ((c3 & 0x3F) << 6) | (c4 & 0x3F);
+			}
+		} else {
+			*out++ = MBFL_BAD_INPUT;
+			continue;
+		}
+
+		int s1 = 0, snd = 0;
+		if (mbfilter_conv_r_map_tbl(s, &s1, emoji_map, n)) {
+			s = convert_emoji(s1, &snd);
+			if (snd) {
+				*out++ = snd;
+			}
+		}
+		*out++ = s;
+	}
+
+	*in_len = e - p;
+	*in = p;
+	return out - buf;
+}
+
+static size_t mb_utf8_docomo_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+{
+	return mb_mobile_utf8_to_wchar(in, in_len, buf, bufsize, state, mbfl_docomo2uni_pua, mbfilter_sjis_emoji_docomo2unicode, 4);
+}
+
+static void mb_wchar_to_utf8_docomo(uint32_t *in, size_t len, mb_convert_buf *buf, bool end)
+{
+	unsigned char *out, *limit;
+	MB_CONVERT_BUF_LOAD(buf, out, limit);
+	MB_CONVERT_BUF_ENSURE(buf, out, limit, len);
+
+	while (len--) {
+		uint32_t w = *in++;
+		unsigned int s = 0;
+		int c1 = 0;
```

Comment:
This seems to only be used in the
```c
			if (s && mbfilter_conv_map_tbl(s, &c1, mbfl_docomo2uni_pua, 4)) {
				w = c1;
			}
```
block of code, with also a type sign conversion happening

Reply:
The reason for this to be `int` is because that's what the tables accessed via `mbfilter_conv_map_tbl` contain.

I think this will be easier to clean up once the legacy code is deleted.

Reply:
Right, I will let you do this then :)